styx-common/src/main/java/com/spotify/styx/api/Middlewares.java

```diff
@@ -79,39 +90,94 @@ private Middlewares() {
     });
   }
 
-  public static Middleware<AsyncHandler<? extends Response<?>>,
-      AsyncHandler<? extends Response<ByteString>>> clientValidator(
+  public static <T> Middleware<AsyncHandler<Response<T>>, AsyncHandler<Response<T>>> clientValidator(
       Supplier<Optional<List<String>>> supplier) {
     return innerHandler -> requestContext -> {
       if (requestContext.request().header("User-Agent")
           .map(header -> supplier.get().orElse(ImmutableList.of()).contains(header))
           .orElse(false)) {
         // TODO: fire some stats
-        return CompletableFuture
-            .completedFuture(Response.forStatus(Status.NOT_ACCEPTABLE.withReasonPhrase(
+        return
+            completedFuture(Response.forStatus(Status.NOT_ACCEPTABLE.withReasonPhrase(
                 "blacklisted client version, please upgrade")));
       } else {
-        // noinspection unchecked
-        return (CompletionStage<Response<ByteString>>) innerHandler.invoke(requestContext);
+        return innerHandler.invoke(requestContext);
       }
     };
   }
 
-  public static Middleware<AsyncHandler<? extends Response<?>>,
-      AsyncHandler<? extends Response<ByteString>>> auditLogging() {
+  private static final GoogleIdTokenVerifier GOOGLE_ID_TOKEN_VERIFIER;
+
+  static {
+    final NetHttpTransport transport;
+    try {
+      transport = GoogleNetHttpTransport.newTrustedTransport();
+    } catch (GeneralSecurityException | IOException e) {
+      throw new RuntimeException(e);
+    }
+    GOOGLE_ID_TOKEN_VERIFIER = new GoogleIdTokenVerifier
+        .Builder(transport, Utils.getDefaultJsonFactory())
+        .build();
+  }
+
+  public interface AuthContext {
+    Optional<GoogleIdToken> user();
+  }
+
+  public interface Authenticated<T> extends Function<AuthContext, T> {}
+  public interface Requested<T> extends Function<RequestContext, T> {}
+
+  public static <T> Middleware<Requested<Authenticated<T>>, AsyncHandler<Response<ByteString>>> authed() {
+    return ar -> jsonAsync().apply(requestContext -> {
+      final T payload = ar
+          .apply(requestContext)
+          .apply(auth(requestContext));
+      return completedFuture(Response.forPayload(payload));
+    });
+  }
+
+  public static <T> Middleware<AsyncHandler<Response<T>>, AsyncHandler<Response<T>>> authValidator() {
+    return h -> rc -> {
+      if (!auth(rc).user().isPresent()) {
+        return completedFuture(Response.forStatus(Status.UNAUTHORIZED));
+      }
+
+      return h.invoke(rc);
+    };
+  }
+
+  public static AuthContext auth(RequestContext requestContext) {
+    return () -> requestContext.request()
+        .header(HttpHeaders.AUTHORIZATION)
+        .filter(s -> s.startsWith(BEARER_PREFIX))
+        .map(s -> s.substring(BEARER_PREFIX.length()))
+        .map(Middlewares::verifyIdToken)
+        .filter(Objects::nonNull);
+  }
+
+  private static GoogleIdToken verifyIdToken(final String s) {
+    try {
+      return GOOGLE_ID_TOKEN_VERIFIER.verify(s);
```

Comment:
What does that mean for key revocations? Is it acceptable to have offline/local verification only? Perhaps it could be enforced on the mutating (POST, DELETE etc) APIs but local validation could be fine on non-mutating (HEAD, GET)?